File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/FluidGridHandler.java
package com.refinedmods.refinedstorage.apiimpl.network.grid.handler;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.api.autocrafting.task.CalculationResultType;
import com.refinedmods.refinedstorage.api.autocrafting.task.ICalculationResult;
import com.refinedmods.refinedstorage.api.network.INetwork;
import com.refinedmods.refinedstorage.api.network.grid.handler.IFluidGridHandler;
import com.refinedmods.refinedstorage.api.network.security.Permission;
import com.refinedmods.refinedstorage.api.util.Action;
import com.refinedmods.refinedstorage.apiimpl.autocrafting.preview.ErrorCraftingPreviewElement;
import com.refinedmods.refinedstorage.network.grid.GridCraftingPreviewResponseMessage;
import com.refinedmods.refinedstorage.network.grid.GridCraftingStartResponseMessage;
import com.refinedmods.refinedstorage.util.NetworkUtils;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.Containers;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.common.capabilities.ForgeCapabilities;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.FluidType;
import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
import net.minecraftforge.fluids.capability.IFluidHandler;
import org.apache.commons.lang3.tuple.Pair;

import javax.annotation.Nonnull;
import java.util.Collections;
import java.util.UUID;

public class FluidGridHandler implements IFluidGridHandler {
    private final INetwork network;

    public FluidGridHandler(INetwork network) {
        this.network = network;
    }

    @Override
    public void onExtract(ServerPlayer player, UUID id, boolean shift) {
        FluidStack stack = network.getFluidStorageCache().getList().get(id);

        if (stack == null || stack.getAmount() < FluidType.BUCKET_VOLUME || !network.getSecurityManager().hasPermission(Permission.EXTRACT, player) || !network.canRun()) {
            return;
        }

        NetworkUtils.extractBucketFromPlayerInventoryOrNetwork(player, network, bucket -> bucket.getCapability(ForgeCapabilities.FLUID_HANDLER_ITEM, null).ifPresent(fluidHandler -> {
            network.getFluidStorageTracker().changed(player, stack.copy());

            FluidStack extracted = network.extractFluid(stack, FluidType.BUCKET_VOLUME, Action.PERFORM);

            fluidHandler.fill(extracted, IFluidHandler.FluidAction.EXECUTE);

            if (shift) {
                if (!player.getInventory().add(fluidHandler.getContainer().copy())) {
                    Containers.dropItemStack(player.getCommandSenderWorld(), player.getX(), player.getY(), player.getZ(), fluidHandler.getContainer());
                }
            } else {
                player.containerMenu.setCarried(fluidHandler.getContainer());
            }

            network.getNetworkItemManager().drainEnergy(player, RS.SERVER_CONFIG.getWirelessFluidGrid().getExtractUsage());
        }));
    }

    @Override
    @Nonnull
    public ItemStack onInsert(ServerPlayer player, ItemStack container) {
        if (!network.getSecurityManager().hasPermission(Permission.INSERT, player) || !network.canRun()) {
            return container;
        }

        Pair<ItemStack, FluidStack> result = StackUtils.getFluid(container, true);

        if (!result.getValue().isEmpty() && network.insertFluid(result.getValue(), result.getValue().getAmount(), Action.SIMULATE).isEmpty()) {
            network.getFluidStorageTracker().changed(player, result.getValue().copy());

            result = StackUtils.getFluid(container, false);

            network.insertFluid(result.getValue(), result.getValue().getAmount(), Action.PERFORM);

            network.getNetworkItemManager().drainEnergy(player, RS.SERVER_CONFIG.getWirelessFluidGrid().getInsertUsage());

            return result.getLeft();
        }

        return container;
    }

    @Override
    public void onInsertHeldContainer(ServerPlayer player) {
        player.containerMenu.setCarried(onInsert(player, player.containerMenu.getCarried()));
    }

    @Override
    public void onCraftingPreviewRequested(ServerPlayer player, UUID id, int quantity, boolean noPreview) {
        if (!network.getSecurityManager().hasPermission(Permission.AUTOCRAFTING, player)) {
            return;
        }

        FluidStack stack = network.getFluidStorageCache().getCraftablesList().get(id);

        if (stack != null) {
            ICalculationResult result = network.getCraftingManager().create(stack, quantity);
            if (result == null) {
                return;
            }

            if (!result.isOk() && result.getType() != CalculationResultType.MISSING) {
                RS.NETWORK_HANDLER.sendTo(
                    player,
                    new GridCraftingPreviewResponseMessage(
                        Collections.singletonList(new ErrorCraftingPreviewElement(result.getType(), result.getRecursedPattern() == null ? ItemStack.EMPTY : result.getRecursedPattern().getStack())),
                        id,
                        quantity,
                        true
                    )
                );
            } else if (result.isOk() && noPreview) {
                network.getCraftingManager().start(result.getTask());

                RS.NETWORK_HANDLER.sendTo(player, new GridCraftingStartResponseMessage());
            } else {
                RS.NETWORK_HANDLER.sendTo(
                    player,
                    new GridCraftingPreviewResponseMessage(
                        result.getPreviewElements(),
                        id,
                        quantity,
                        true
                    )
                );
            }
        }
    }

    @Override
    public void onCraftingRequested(ServerPlayer player, UUID id, int quantity) {
        if (quantity <= 0 || !network.getSecurityManager().hasPermission(Permission.AUTOCRAFTING, player)) {
            return;
        }

        FluidStack stack = network.getFluidStorageCache().getCraftablesList().get(id);

        if (stack != null) {
            ICalculationResult result = network.getCraftingManager().create(stack, quantity);
            if (result.isOk()) {
                network.getCraftingManager().start(result.getTask());
            }
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/FluidGridHandler.kt
package com.refinedmods.refinedstorage.apiimpl.network.grid.handler

import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.api.autocrafting.task.CalculationResultType
import com.refinedmods.refinedstorage.api.autocrafting.task.ICalculationResult
import com.refinedmods.refinedstorage.api.network.INetwork
import com.refinedmods.refinedstorage.api.network.grid.handler.IFluidGridHandler
import com.refinedmods.refinedstorage.api.network.security.Permission
import com.refinedmods.refinedstorage.api.util.Action
import com.refinedmods.refinedstorage.apiimpl.autocrafting.preview.ErrorCraftingPreviewElement
import com.refinedmods.refinedstorage.network.grid.GridCraftingPreviewResponseMessage
import com.refinedmods.refinedstorage.network.grid.GridCraftingStartResponseMessage
import com.refinedmods.refinedstorage.util.NetworkUtils
import com.refinedmods.refinedstorage.util.StackUtils.getFluid
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.inventory.InventoryHelper
import net.minecraft.item.ItemStack
import net.minecraftforge.fluids.FluidAttributes
import net.minecraftforge.fluids.FluidInstance
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.minecraftforge.fluids.capability.IFluidHandler
import org.apache.commons.lang3.tuple.Pair
import java.util.*


class FluidGridHandler(private val network: INetwork) : IFluidGridHandler {
    override fun onExtract(player: ServerPlayerEntity?, id: UUID?, shift: Boolean) {
        val stack: FluidInstance? = network.fluidStorageCache!!.getList()!![id]
        if (stack == null || stack.getAmount() < FluidAttributes.BUCKET_VOLUME || !network.securityManager!!.hasPermission(Permission.EXTRACT, player)) {
            return
        }
        NetworkUtils.extractBucketFromPlayerInventoryOrNetwork(player, network) { bucket: ItemStack ->
            bucket.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY, null).ifPresent({ fluidHandler ->
                network.fluidStorageTracker!!.changed(player, stack.copy())
                val extracted: FluidInstance? = network.extractFluid(stack, FluidAttributes.BUCKET_VOLUME, Action.PERFORM)
                fluidHandler.fill(extracted, IFluidHandler.FluidAction.EXECUTE)
                if (shift) {
                    if (!player.inventory.addItemStackToInventory(fluidHandler.getContainer().copy())) {
                        InventoryHelper.spawnItemStack(player.getEntityWorld(), player.getPosX(), player.getPosY(), player.getPosZ(), fluidHandler.getContainer())
                    }
                } else {
                    player.inventory.setItemStack(fluidHandler.getContainer())
                    player.updateHeldItem()
                }
                network.networkItemManager!!.drainEnergy(player, RS.SERVER_CONFIG.wirelessFluidGrid.extractUsage)
            })
        }
    }

    @Nonnull
    override fun onInsert(player: ServerPlayerEntity?, container: ItemStack?): ItemStack? {
        if (!network.securityManager!!.hasPermission(Permission.INSERT, player)) {
            return container
        }
        var result: Pair<ItemStack?, FluidInstance> = getFluid(container!!, true)
        if (!result.value.isEmpty() && network.insertFluid(result.value, result.value.getAmount(), Action.SIMULATE).isEmpty()) {
            network.fluidStorageTracker!!.changed(player, result.value.copy())
            result = getFluid(container, false)
            network.insertFluid(result.value, result.value.getAmount(), Action.PERFORM)
            network.networkItemManager!!.drainEnergy(player, RS.SERVER_CONFIG.wirelessFluidGrid.insertUsage)
            return result.left
        }
        return container
    }

    override fun onInsertHeldContainer(player: ServerPlayerEntity?) {
        player.inventory.setItemStack(onInsert(player, player.inventory.getItemStack()))
        player.updateHeldItem()
    }

    override fun onCraftingPreviewRequested(player: ServerPlayerEntity?, id: UUID?, quantity: Int, noPreview: Boolean) {
        if (!network.securityManager!!.hasPermission(Permission.AUTOCRAFTING, player)) {
            return
        }
        val stack: FluidInstance? = network.fluidStorageCache!!.getCraftablesList()!![id]
        if (stack != null) {
            val calculationThread = Thread(label@ Runnable {
                val result: ICalculationResult = network.craftingManager.create(stack, quantity) ?: return@label
                if (!result.isOk() && result.getType() !== CalculationResultType.MISSING) {
                    RS.NETWORK_HANDLER.sendTo(
                            player,
                            GridCraftingPreviewResponseMessage(listOf(ErrorCraftingPreviewElement(result.getType(), if (result.getRecursedPattern() == null) ItemStack.EMPTY else result.getRecursedPattern()!!.getStack())),
                                    id,
                                    quantity,
                                    true
                            )
                    )
                } else if (result.isOk() && noPreview) {
                    network.craftingManager.start(result.getTask())
                    RS.NETWORK_HANDLER.sendTo(player, GridCraftingStartResponseMessage())
                } else {
                    RS.NETWORK_HANDLER.sendTo(
                            player,
                            GridCraftingPreviewResponseMessage(
                                    result.getPreviewElements(),
                                    id,
                                    quantity,
                                    true
                            )
                    )
                }
            }, "RS crafting preview calculation")
            calculationThread.start()
        }
    }

    override fun onCraftingRequested(player: ServerPlayerEntity?, id: UUID?, quantity: Int) {
        if (quantity <= 0 || !network.securityManager!!.hasPermission(Permission.AUTOCRAFTING, player)) {
            return
        }
        val stack: FluidInstance? = network.fluidStorageCache!!.getCraftablesList()!![id]
        if (stack != null) {
            val result: ICalculationResult = network.craftingManager.create(stack, quantity)
            if (result.isOk()) {
                network.craftingManager.start(result.getTask())
            }
        }
    }
}